enhancements/baremetal/baremetal-provisioning-optional.md

```diff
@@ -0,0 +1,179 @@
+---
+title: baremetal-provisioning-optional
+authors:
+  - "@stbenjam"
+reviewers:
+  - "@dhellman"
+  - "@hardys"
+  - "@kirankt"
+  - "@sadasu"
+approvers:
+  - "@hardys"
+creation-date: 2020-05-21
+last-updated: 2020-06-23
+status: provisional
+see-also:
+  - "/enhancements/baremetal/baremetal-provisioning-config.md"
+  - "/enhancements/installer/connected-assisted-installer.md"
+---
+
+# Baremetal Optional Provisioning Network
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Today the baremetal IPI platform takes a "batteries included" approach
+that has a dedicated provisioning network, HTTP caches, DHCP and TFTP
+servers, fully managed provisioning with both PXE and virtual media, and
+some level of hardware management.
+
+In some situations it may be possible to remove the requirement for a
+provisioning network entirely, in particular in the case where
+virtual media or the assisted install mechanism is used as the
+provisioning method.
+
+Previously, we added a flag to the installer that lets a user disable
+DHCP, this enhancement builds on that by making a configurable
+provisioning network profile to allow for managed, unmanaged, and
+disabled configurations.
+
+## Motivation
+
+In order to reduce the number of boolean flags in the baremetal
+platform, the various options for the provisioning network will be
+consolidated into a single enum field.
+
+### Goals
+
+The goal of this enhancement is to allow installation of a baremetal IPI
+cluster without any provisioning network dependency.
+
+### Non-Goals
+
+We do not intend to remove the static provisioning IP's at this time,
+but instead they will need to be IP's available in the external network.
+
+## Proposal
+
+A new `ProvisioningNetwork` option will be added to the baremetal
+installer platform, that features an enum of possible values:
+
+  - `managed` (default): Fully managed provisioning networking including
+     DHCP, TFTP, etc.
+
+  - `unmanaged`: Provisioning network is still present and used, but
+     user is responsible for managing DHCP. Virtual media provisioning
+     is reccomended, but PXE is still available if required.
+
+  - `disabled`: Provisioning network is fully disabled. User may only do
+     virtual media based provisioning, or bring up the cluster using
+     assisted installation. If using power management, BMC's must be
+     accessible from the machine networks. User must provide 2 IP's on
+     the external network that are used for the provisioning services.
+
+The same field will be added to the Provisioning CRD, with the
+provisioningDHCPExternal field removed.
```

Comment:
Removing the existing field will take a couple of releases. In the mean time, we will need to declare which field takes precedence over the other and support not having the `ProvisioningNetwork` option set at all.

Reply:
What is the benefit of removing it?  Is it worth breaking backwards compatibility instead of just dropping it from docs?

Can you also add a reference to the current definition of this CRD?

Reply:
Hm ok, I'll update this section. We could leave it around forever I suppose, but undocumented.

Reply:
Note that the Provisioning CRD was added purely to enable the installer to pass data to the MAO ref https://github.com/openshift/machine-api-operator/pull/460 - we had a bunch of discussion on https://github.com/openshift/api/pull/540 before that, and one of the reasons it wasn't added to openshift/api was avoid publishing any "external" interface.

So IMO it's probably fine to deprecate then later remove the existing option, given that it's highly likely that the only user of this interface is the installer?

Reply:
ACK thanks for clarifying

Reply:
I clarified this, please have a look to see if it's clear.